src/logger.c

```diff
@@ -136,24 +137,37 @@ int log_lock(int fd, int cmd, int type, const char *name,
 	        : cmd == F_UNLCK ? "F_UNLCK" : "");
 #endif
 
+	uint64_t lock_start = 0;
+	int warned = 0;
+	int retry = 0;
+
 	lock.l_type = type;
 	while (fcntl(fd, cmd, &lock)) {
-		if (errno == EAGAIN) {
-			static int warned;
+		if (errno == EAGAIN || errno == ENOLCK) {
 			struct timeval t;
+			unsigned int warn_bit = (errno == EAGAIN ? 1 : 2);
 
 			if (cmd == F_SETLK)
 				return -1;
 
-			if (!warned++) {
-				log_event("Got EAGAIN despite F_SETLKW (only logged once per node) %s:%d %s", source_file, line, function);
-				fprintf(stderr, "Node %d: File locking apparently exhausted, check ulimits and any NFS server limits. This is recoverable but will harm performance (muting further of these messages from same node)\n", NODE);
+			if (!(warned & warn_bit)) {
+				if (!lock_start)
+					lock_start = john_get_nano();
+
+				log_event("- Got %s despite F_SETLKW on %s, %s() %s:%d", errno == EAGAIN ? "EAGAIN" : "ENOLCK", name, function, source_file, line);
+				if (options.node_count)
+					fprintf(stderr, "%u: ", NODE);
+				fprintf(stderr, "File locking apparently exhausted (%s), retrying. Check ulimits and any NFS server limits\n", strerror(errno));
 				srand(NODE);
+
+				warned |= warn_bit;
 			}
 
 			/* Sleep for a random time of max. ~260 ms */
 			t.tv_sec = 0; t.tv_usec = (rand() & 1023) << 8;
```

Comment:
Oh, I see by using `srand(NODE);` this code tries to have different delays for different nodes to avoid clashes. However, besides the issue I just pointed out, this is also not great because this PRNG is shared with other uses we have in the tree:

```
./logger.c:			t.tv_sec = 0; t.tv_usec = (rand() & 1023) << 8;
./dynamic_compiler.c:		int j = rand() % 16;
./oaes_lib_plug.c:		_key->data[_i] = (uint8_t) rand( _ctx->rctx );
./oaes_lib_plug.c:		_key->data[_i] = (uint8_t) rand();
./oaes_lib_plug.c:					_ctx->iv[_i] = (uint8_t) rand( _ctx->rctx );
./oaes_lib_plug.c:					_ctx->iv[_i] = (uint8_t) rand();
```

```
./logger.c:				srand(NODE);
./oaes_lib_plug.c:	srand(0); // XXX
```

We could also want to check those other uses and maybe get rid of (some of) them.

Reply:
> We could also want to check those other uses and maybe get rid of (some of) them.

Got rid of those in `oaes_lib_plug.c` since they didn't affect anything in our usage anyway.